File: clients/kotlin_graal/src/main/kotlin/ai_cup_22/model/Sound.kt
package ai_cup_22.model

import ai_cup_22.util.StreamUtil

/**
 * Sound heard by one of your units
 */
class Sound {
    /**
     * Sound type index (starting with 0)
     */
    var typeIndex: Int
    /**
     * Id of unit that heard this sound
     */
    var unitId: Int
    /**
     * Position where sound was heard (different from sound source position)
     */
    var position: ai_cup_22.model.Vec2

    constructor(typeIndex: Int, unitId: Int, position: ai_cup_22.model.Vec2) {
        this.typeIndex = typeIndex
        this.unitId = unitId
        this.position = position
    }

    /**
     * Write Sound to output stream
     */
    @Throws(java.io.IOException::class)
    fun writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, typeIndex)
        StreamUtil.writeInt(stream, unitId)
        position.writeTo(stream)
    }

    /**
     * Get string representation of Sound
     */
    override fun toString(): String {
        var stringBuilder = StringBuilder("Sound { ")
        stringBuilder.append("typeIndex: ")
        stringBuilder.append(typeIndex)
        stringBuilder.append(", ")
        stringBuilder.append("unitId: ")
        stringBuilder.append(unitId)
        stringBuilder.append(", ")
        stringBuilder.append("position: ")
        stringBuilder.append(position)
        stringBuilder.append(" }")
        return stringBuilder.toString()
    }

    companion object {
        /**
         * Read Sound from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): Sound {
            var typeIndex: Int
            typeIndex = StreamUtil.readInt(stream)
            var unitId: Int
            unitId = StreamUtil.readInt(stream)
            var position: ai_cup_22.model.Vec2
            position = ai_cup_22.model.Vec2.readFrom(stream)
            return Sound(typeIndex, unitId, position)
        }
    }
}

File: clients/swift/Sources/AiCup22/Model/Sound.swift
/// Sound heard by one of your units
public struct Sound {
    /// Sound type index (starting with 0)
    let typeIndex: Int32

    /// Id of unit that heard this sound
    let unitId: Int32

    /// Position where sound was heard (different from sound source position)
    let position: Vec2

    /// Read Sound from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> Sound {
        var typeIndex: Int32
        typeIndex = stream.readInt32()
        var unitId: Int32
        unitId = stream.readInt32()
        var position: Vec2
        position = Vec2.readFrom(stream)
        return Sound(typeIndex: typeIndex, unitId: unitId, position: position)
    }

    /// Write Sound to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        stream.writeInt32(typeIndex)
        stream.writeInt32(unitId)
        position.writeTo(stream)
    }
}
